pkg/compute/kube/client.go

```diff
@@ -406,3 +387,125 @@ func DetectIngressBackendConflicts(x policyv1alpha1.IngressBackend, y policyv1al
+				Name:      pod.Spec.ServiceAccountName,
+				Namespace: pod.Namespace, // ServiceAccount must belong to the same namespace as the pod
+			}
+			svcAccountsSet.Add(podSvcAccount)
+		}
+	}
+
+	for svcAcc := range svcAccountsSet.Iter() {
+		identities = append(identities, svcAcc.(identity.K8sServiceAccount).ToServiceIdentity())
+	}
+	return identities, nil
+}
+
+// GetMeshService returns the service.MeshService corresponding to the Port used by clients
+// to communicate with it.
+func (c *client) GetMeshService(name, namespace string, port uint16) (service.MeshService, error) {
+	v1Svc := c.kubeController.GetService(name, namespace)
+	if v1Svc == nil {
+		return service.MeshService{}, errServiceNotFound
+	}
+	for _, svc := range c.serviceToMeshServices(*v1Svc) {
+		if svc.Port == port {
+			return svc, nil
+		}
+	}
+	return service.MeshService{}, fmt.Errorf("service %s/%s does not have a port %d", namespace, name, port)
+}
+
+// serviceToMeshServices translates a k8s service with one or more ports to one or more
+// MeshService objects per port.
+func (c *client) serviceToMeshServices(svc corev1.Service) []service.MeshService {
+	var meshServices []service.MeshService
+
+	for _, portSpec := range svc.Spec.Ports {
+		meshSvc := service.MeshService{
+			Namespace: svc.Namespace,
+			Name:      svc.Name,
+			Port:      uint16(portSpec.Port),
+		}
+
+		// attempt to parse protocol from port name
+		// Order of Preference is:
+		// 1. port.appProtocol field
+		// 2. protocol prefixed to port name (e.g. tcp-my-port)
+		// 3. default to http
+		protocol := constants.ProtocolHTTP
+		for _, p := range constants.SupportedProtocolsInMesh {
+			if strings.HasPrefix(portSpec.Name, p+"-") {
+				protocol = p
+				break
+			}
+		}
+
+		// use port.appProtocol if specified, else use port protocol
+		meshSvc.Protocol = pointer.StringDeref(portSpec.AppProtocol, protocol)
+
+		// The endpoints for the kubernetes service carry information that allows
+		// us to retrieve the TargetPort for the MeshService.
+		endpoints, _ := c.kubeController.GetEndpoints(svc.Name, svc.Namespace)
+		if endpoints != nil {
+			meshSvc.TargetPort = GetTargetPortFromEndpoints(portSpec.Name, *endpoints)
+		} else {
+			log.Warn().Msgf("k8s service %s/%s does not have endpoints but is being represented as a MeshService", svc.Namespace, svc.Name)
+		}
+		if !IsHeadlessService(svc) || endpoints == nil {
+			meshServices = append(meshServices, meshSvc)
+			continue
+		}
+
+		for _, subset := range endpoints.Subsets {
+			for _, address := range subset.Addresses {
+				if address.Hostname == "" {
+					continue
+				}
+				meshServices = append(meshServices, service.MeshService{
+					Namespace:  svc.Namespace,
+					Name:       svc.Name,
+					Subdomain:  address.Hostname,
+					Port:       meshSvc.Port,
+					TargetPort: meshSvc.TargetPort,
+					Protocol:   meshSvc.Protocol,
+				})
+			}
+		}
+	}
+	return meshServices
+}
+
+func (c *client) ListNamespaces() ([]string, error) {
```

Comment:
this looks like it is used by the debug handler endpoint which already has a reference to the [kubecontroller](https://github.com/steeling/osm/blob/2e96617b4abe8daed6b8c963f75333f59194c4f7/pkg/debugger/server.go#L52). Is this needed by other components or could it be dropped?

Reply:
ya i have further plans to remove the kubecontroller from the debugger.

The reason I put it here is that `ListNamespaces` in the k8s/client.go should return a k8s namespace object, while the OSM domain representation of that is simply a string, denoting that it belongs in the compute/client